content/docs/capabilities/mcp.mdx

```diff
@@ -0,0 +1,587 @@
+
+**Characteristics:**
+
+- Automatically refreshed by Pomerium
+- Securely stored and managed
+- Used for upstream API authentication
+- Invisible to external clients
+
+## Configuration Options
+
+MCP integration can be configured through route-level settings. For general route configuration guidance, see the [Routes configuration reference](/docs/reference/routes).
+
+### MCP Server Configuration
+
+For routes hosting MCP servers:
+
+```yaml
+mcp:
+  server:
+    # Optional: Configure upstream OAuth2 for services requiring authentication
+    upstream_oauth2:
+      client_id: 'your-oauth-client-id'
+      client_secret: 'your-oauth-client-secret'
+      endpoint:
+        auth_url: 'https://provider.com/oauth/authorize'
+        token_url: 'https://provider.com/oauth/token'
+        auth_style: 'header' # or "params"
+      scopes: ['scope1', 'scope2']
+
+    # Optional: Maximum request body size (default: 4KiB)
+    max_request_bytes: 1048576
+```
+
+** Tool Call Authorization **
+
+When using [Pomerium Policy Language (PPL)](/docs/internals/ppl) with MCP routes, you can control access to specific tools, in addition to the standard route policies. For example, to allow only certain users to call a specific MCP tool:
+
+```yaml
+policy:
+  allow:
+    and:
+      - email:
+          is: analyst@company.com
+      - mcp_tool:
+          is: database_query
+```
+
+For more information on route-level policies, see [Route Policy Configuration](/docs/reference/routes/policy). For advanced policy examples, see [Policy-Based Tool Access Control](#policy-based-tool-access-control).
+
+### MCP Client Configuration
+
+For applications that consume MCP servers:
+
+```yaml
+mcp:
+  # Provides the application with external tokens for MCP server access.
+  # the brackets are significant, they indicate that this is an MCP client route
+  client: {}
+```
+
+For Client routes, Pomerium exposes additional endpoints to manage MCP server discovery and upstream OAuth flows. See [Building MCP-Enabled Applications](#3-building-mcp-enabled-applications) for architectural details.
+
+#### Listing Available MCP Servers
+
+Applications can discover available MCP servers by making a request to the `/.pomerium/mcp/routes` endpoint:
+
+```http
+GET https://your-app.domain.com/.pomerium/mcp/routes
+Authorization: Bearer <external-token>
+Accept: application/json
+```
+
+**Response:**
+
+```json
+{
+  "servers": [
+    {
+      "name": "Database Server",
+      "url": "https://db-mcp.your-domain.com",
+      "connected": true
+    },
+    {
+      "name": "GitHub Server",
+      "url": "https://github-mcp.your-domain.com",
+      "connected": false
+    }
+  ]
+}
+```
+
+The `connected` field indicates whether the user has completed all required upstream OAuth flows.
+
+#### Initiating Upstream OAuth
+
+If a server shows `connected: false`, redirect users to complete upstream authentication:
```

Comment:
Maybe for later, but a section about a JWT secret for deploying to e.g. Netlify or Vercel?